src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -0,0 +1,467 @@
+
+#include <Storages/StorageFactory.h>
+#include <Storages/Cache/SchemaCache.h>
+#include <Storages/VirtualColumnUtils.h>
+#include <Storages/ObjectStorage/Utils.h>
+#include <Storages/NamedCollectionsHelpers.h>
+#include <Storages/ObjectStorage/StorageObjectStorage.h>
+#include <Storages/ObjectStorage/StorageObjectStorageSink.h>
+#include <Storages/ObjectStorage/StorageObjectStorageSource.h>
+#include <Storages/ObjectStorage/ReadBufferIterator.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int DATABASE_ACCESS_DENIED;
+    extern const int NOT_IMPLEMENTED;
+    extern const int LOGICAL_ERROR;
+}
+
+StorageObjectStorage::StorageObjectStorage(
+    ConfigurationPtr configuration_,
+    ObjectStoragePtr object_storage_,
+    ContextPtr context,
+    const StorageID & table_id_,
+    const ColumnsDescription & columns_,
+    const ConstraintsDescription & constraints_,
+    const String & comment,
+    std::optional<FormatSettings> format_settings_,
+    bool distributed_processing_,
+    ASTPtr partition_by_)
+    : IStorage(table_id_)
+    , configuration(configuration_)
+    , object_storage(object_storage_)
+    , format_settings(format_settings_)
+    , partition_by(partition_by_)
+    , distributed_processing(distributed_processing_)
+    , log(getLogger(fmt::format("Storage{}({})", configuration->getEngineName(), table_id_.getFullTableName())))
+{
+    ColumnsDescription columns{columns_};
+    resolveSchemaAndFormat(columns, configuration->format, object_storage, configuration, format_settings, context);
+    configuration->check(context);
+
+    StorageInMemoryMetadata metadata;
+    metadata.setColumns(columns);
+    metadata.setConstraints(constraints_);
+    metadata.setComment(comment);
+
+    StoredObjects objects;
+    for (const auto & key : configuration->getPaths())
+        objects.emplace_back(key);
+
+    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns()));
+    setInMemoryMetadata(metadata);
+}
+
+String StorageObjectStorage::getName() const
+{
+    return configuration->getEngineName();
+}
+
+bool StorageObjectStorage::prefersLargeBlocks() const
+{
+    return FormatFactory::instance().checkIfOutputFormatPrefersLargeBlocks(configuration->format);
+}
+
+bool StorageObjectStorage::parallelizeOutputAfterReading(ContextPtr context) const
+{
+    return FormatFactory::instance().checkParallelizeOutputAfterReading(configuration->format, context);
+}
+
+bool StorageObjectStorage::supportsSubsetOfColumns(const ContextPtr & context) const
+{
+    return FormatFactory::instance().checkIfFormatSupportsSubsetOfColumns(configuration->format, context, format_settings);
+}
+
+void StorageObjectStorage::updateConfiguration(ContextPtr context)
+{
+    /// FIXME: we should be able to update everything apart from client if static_configuration == true.
+    if (!configuration->isStaticConfiguration())
+        object_storage->applyNewSettings(context->getConfigRef(), "s3.", context);
```

Comment:
Why prefix is `s3.`? Should it be `configuration->getTypeName()` or smth like this?